### PR TITLE
Allow scenarios for logged in users

### PR DESCRIPTION
**Pull request by jdlrobson**

### CODE REVIEW

File: src/engine-scripts/puppet/loginReload.js [hunk 1]
```diff
@@ -0,0 +1,10 @@
+const loginReload = async (page, scenario) => {
+    await page.goto(scenario.loginUrl);
```
Comment (nicholasray) [inline src/engine-scripts/puppet/loginReload.js:2]: Was config.js scenarios edited to include the loginUrl?

Comment (jdlrobson) [reply src/engine-scripts/puppet/loginReload.js:2]: Ack. I messed that up.

Comment (jdlrobson) [reply src/engine-scripts/puppet/loginReload.js:2]: Fixed in latest patchset!

**Comment by jdlrobson** [comment]
Latest patch seems more stable - using onBefore instead of onReady.

**Comment by nicholasray** [comment]
Thank you for working on this.

I'm still investigating the cause, but I've noticed the following error message 2 out of 3 runs of `./pixel.js reference` with this change:

```
Puppeteer encountered an error while running scenario "Special:BlankPage with user menu open (#vector-2022, #logged-in, #userMenu-open)"
TimeoutError: Navigation timeout of 60000 ms exceeded
```

File: src/engine-scripts/puppet/scroll.js [hunk 2]
```diff
@@ -0,0 +1,5 @@
+module.exports = (page) => {
+	page.evaluate( () => {
```
Comment (nicholasray) [inline src/engine-scripts/puppet/scroll.js:2]: Does this need await?

**Comment by jdlrobson** [comment]
Was this with the version of the patch using onReady? I was getting those errors with that but not onBefore

File: src/engine-scripts/puppet/scroll.js (hunk 2 above)
Comment (jdlrobson) [reply src/engine-scripts/puppet/scroll.js:2]: I don't know why but i was getting a timeout with this, presumably because it never resolved

**Comment by nicholasray** [comment]
@jdlrobson Yes, it's with commit 91b478625deaf9f5bb9ad74dadd3320484626d54  . But it might be due to the await question I posted above

File: src/engine-scripts/puppet/scroll.js (hunk 2 above)
Comment (nicholasray) [reply src/engine-scripts/puppet/scroll.js:2]: It happens infrequently but definitely happens for me:
Screen Shot 2022-04-18 at 2 57 13 PM

Not sure why this is yet as the adding async/await didn't seem to matter.

The other issue I'm seeing is the sticky header tests aren't showing the sticky header for me:
MediaWiki_Test_sticky_header_logged-in_scroll_0_viewport_1_tablet

**Comment by jdlrobson** [comment]
Sticky header shouldn't be showing on tablet view. Are you not seeing it on the desktop breakpoint?

**Comment by nicholasray** [comment]
Ah good point. I see this at the "desktop" (not to be confused with the "desktop-wide" breakpoint which shows the sticky header. Presumably at this breakpoint we shouldn't show the sticky header but it looks like our latest TOC styles are not equipped to handle this right now as there is a large gap where the sticky header would otherwise go:

MediaWiki_Test_sticky_header_logged-in_scroll_0_viewport_2_desktop

File: src/engine-scripts/puppet/scroll.js (hunk 2 above)
Comment (jdlrobson) [reply src/engine-scripts/puppet/scroll.js]: Okay, FINALLY worked this one out. I needed to return true in page.evaluate

Pushed new patch.

File: src/engine-scripts/puppet/loginReload.js [hunk 3]
```diff
@@ -0,0 +1,11 @@
+const loginReload = async (page, scenario) => {
+    const INPUT_SELECTOR = 'input[name="wpName"]';
+    await page.goto(scenario.loginUrl);
+    await page.waitForSelector(INPUT_SELECTOR);
+    await page.type(INPUT_SELECTOR, scenario.username);
+    await page.type('input[name="wpPassword"]', scenario.password);
+    await page.click('button[type="submit"]');
```
Comment (nicholasray) [inline src/engine-scripts/puppet/loginReload.js:7]:
```suggestion
    page.click('button[type="submit"]');
```

From debugging the timeouts I've been getting, it looks like having this `await` might be problematic. There might be a race condition happening where by the time this promise has resolved, the navigation is in the middle or at the end leading to L8 timing out. When removing this await, I haven't received a timeout yet, but I'm still checking

Comment (nicholasray) [reply src/engine-scripts/puppet/loginReload.js:7]: Or actually maybe using something like as described at https://github.com/puppeteer/puppeteer/issues/1412#issuecomment-345357063

```
await Promise.all([
  page.waitForNavigation(),
  page.click('a')
]);
```

**Comment by nicholasray** [comment]
@jdlrobson I tried an experiment with the last commit to see how much faster it would be to use an already retrieved cookie without an expiration (from my understanding). The tests are consistently 10-15 seconds faster compared to the previous commit which uses the login page. I really like the improvement in speed. What do you think?

**Review by jdlrobson — COMMENTED** [review]
I guess since we're currently storing the password/username and there's no public instance of this, this should be okay. It would break however if the internals of how we log in change, but I guess this is fine for now.

**Pull request merged**